content/blog/2020-10-tokio-0-3.md

```diff
@@ -0,0 +1,138 @@
+
+- `dns`, tcp`, `udp` and `uds` -> `net`
+- `rt-util` and `rt-core` -> `rt`
+- `rt-threaded` -> `rt-multi-threaded`
+
+## Runtime and Builder refactor
+
+The Tokio team has removed  `threaded_scheduler` and `basic_scheduler` from the public API. Instead, the runtime `Builder` type introduces two constructors: one constructor for a multi-threaded scheduler and one for a current thread scheduler. These constructors provide a misuse-resistant form of building a runtime that *does not* require fiddling with feature flags and builder methods. Tokio will no longer choose the runtime variant based on feature flags when using `Runtime::new` but instead will default to the multi-threaded runtime and is only available under the `rt` feature flag. In addition, we’ve renamed the `core_threads` builder method to the more accurate `worker_threads`. 
+
+We have also refactored the runtime module. Tokio 0.2 exposed two types to interact with the Tokio runtime: `Runtime` and `Handle`. `Runtime` has a `block_on` method that required `&mut self`, whereas `Handle` has a `block_on` method that required `&self`. With Tokio 0.3, we collapsed Runtime and `Handle` into a single `Runtime`. Tokio 0.3’s `Runtime`‘s `block_on` now accepts `&self`, which supports concurrent calls to `block_on`. This also means `Runtime` implements `Send` and `Sync` and can therefore be stored in an `Arc`.
+
+An example configuring a multi-threaded runtime with 6 worker threads:
+
+```rust
+use tokio::runtime::Builder;
+
+let rt = Builder::new_multi_thread()
+    .worker_threads(6)
+    .enable_all()
+    .build()
+    .unwrap();
+
+rt.block_on(async {
+    println!("Hello world!");
+});
+```
+
+## Methods changed to `&self`
+
+We have updated our io types to use an intrusive linked list for storing the wakers. This has allowed us to change many of the methods on types like `TcpStream` and `UdpSocket` to merely take `&self`. This makes it easier to work with our io types across tasks.
+
+## Removal of non-1.0 crates from the public API
+
+In our push for 1.0 we have removed many non-1.0 dependencies from the public API. This includes `bytes` and `mio`. This will allow us to push for a simpler and more stable 1.0 while we continue to innovate in our low-level crates.
+
+## Mio 0.7
+
+We have finally also upgraded `mio` to 0.7 which was originally released in an alpha in December of 2019 and did not make the cut for Tokio 0.2. Since, then `mio` 0.7 has had time to mature and is finally making it into this release of Tokio 0.3. This upgrades some key dependencies like `winapi` from 0.2 to 0.3.
```

Comment:
```suggestion
We have finally also upgraded `mio` to 0.7 which was originally released in an alpha in December of 2019 and did not make the cut for Tokio 0.2. Since then `mio` 0.7 has had time to mature and is finally making it into this release of Tokio 0.3. This upgrades some key dependencies like `winapi` from 0.2 to 0.3.
```